protocolupgrade/snapshot.go

```diff
@@ -0,0 +1,126 @@
+// Copyright (c) 2022 Gobalsky Labs Limited
+//
+// Use of this software is governed by the Business Source License included
+// in the LICENSE file and at https://www.mariadb.com/bsl11.
+//
+// Change Date: 18 months from the later of the date of the first publicly
+// available Distribution of this version of the repository, and 25 June 2022.
+//
+// On the date above, in accordance with the Business Source License, use
+// of this software will be governed by version 3 or later of the GNU General
+// Public License.
+
+package protocolupgrade
+
+import (
+	"context"
+	"encoding/hex"
+	"sort"
+
+	eventspb "code.vegaprotocol.io/protos/vega/events/v1"
+	snappb "code.vegaprotocol.io/protos/vega/snapshot/v1"
+	"code.vegaprotocol.io/vega/types"
+	"github.com/golang/protobuf/proto"
+)
+
+func (e *Engine) Namespace() types.SnapshotNamespace {
+	return types.ProtocolUpgradeSnapshot
+}
+
+func (e *Engine) Keys() []string {
+	return e.hashKeys
+}
+
+func (e *Engine) Stopped() bool {
+	return false
+}
+
+// get the serialised form and hash of the given key.
+func (e *Engine) serialise(k string) ([]byte, error) {
+	events := make([]*eventspb.ProtocolUpgradeEvent, 0, len(e.activeProposals))
+	for _, evt := range e.events {
+		events = append(events, evt)
+	}
+
+	sort.SliceStable(events, func(i, j int) bool {
+		if events[i].VegaReleaseTag == events[j].VegaReleaseTag {
+			if events[i].DataNodeReleaseTag == events[j].DataNodeReleaseTag {
+				return events[i].UpgradeBlockHeight < events[j].UpgradeBlockHeight
+			}
+			return events[i].DataNodeReleaseTag < events[j].DataNodeReleaseTag
+		}
+		return events[i].VegaReleaseTag < events[j].VegaReleaseTag
+	})
```

Comment:
why not just sort over the protocol IDs ids instead? I means your way is probably better as requires only 1 iteration though